### PR TITLE
#5: Setup CI pipeline

### PR DESCRIPTION
**Pull request by heshamMassoud**
1. Sets up travis script. 
2. Uses `com.bmuschko.docker-java-application` gradle plugin for building and pushing docker images.
3. Dynamically set the version of the library according to the release

~~The build is failing due to non-existent credentials for docker hub repo **yet**.~~
Now they are added. Thanks to @anas-aso

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+- rm -fr $HOME/.gradle/caches/*/plugin-resolution/
+cache:
+  directories:
+  - $HOME/.gradle/caches/
+  - $HOME/.gradle/wrapper/
+after_success:
+  - bash <(curl -s https://codecov.io/bash)
+  - ./gradlew shadowJar --daemon || travis_terminate 1
+  - ./gradlew dockerPushImage -Dbuild.version=$TRAVIS_COMMIT --daemon || travis_terminate 1
+  - ./gradlew dockerPushImage -Dbuild.version=$TRAVIS_BRANCH --daemon || travis_terminate 1
```
Comment (butenkor) [inline .travis.yml]: What about created git tags/releases?

**Comment by heshamMassoud** [comment]
> Usage readme is going to be provided later?

@butenkor yes it will come in #4

File: .travis.yml (hunk 1 above)
Comment (heshamMassoud) [reply .travis.yml]: If it's a tag/release TRAVIS_BRANCH will be the name of the tag. :


> `TRAVIS_BRANCH`
for push builds, or builds not triggered by a pull request, this is the name of the branch.
for builds triggered by a pull request this is the name of the branch targeted by the pull request.
for builds triggered by a tag, this is the same as the name of the tag (TRAVIS_TAG).

File: build.gradle [hunk 2]
```diff
@@ -14,8 +14,10 @@ plugins {
     id "com.github.ben-manes.versions" version '0.20.0'
     id 'com.adarshr.test-logger' version '1.6.0'
     id "com.diffplug.gradle.spotless" version '3.16.0'
+    id 'com.bmuschko.docker-java-application' version '4.1.0'
```
Comment (ahmetoz) [inline build.gradle]: it seems the maintainer released the `4.2.0`. Maybe we can use it.

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+##### Full build with tests
+````bash
+./gradlew clean build
+````
+
+### Build 
+ After every successful build on travis, a Docker image is built. Every image is tagged twice:
+ - `$TRAVIS_COMMIT` [travis' environment variable](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables)
+ - `$TRAVIS_BRANCH` [travis' environment variable](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables)
```
Comment (ahmetoz) [inline docs/CONTRIBUTING.md]: why we need for branch tag if we have commit tag?

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+./gradlew spotlessCheck
+````
+
+##### Fix Spotless style violations
+````bash
+./gradlew spotlessApply
+````
+
+It uses the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html) as rules for the style. 
+It is recommended to set it in your IDE auto formatting settings for this project.
```
Comment (ahmetoz) [inline docs/CONTRIBUTING.md]: maybe we can add [this](https://github.com/google/google-java-format) to help how to add ide settings. it is documented here.

File: gradle-scripts/docker.gradle [hunk 5]
```diff
@@ -0,0 +1,13 @@
+docker {
+    javaApplication {
+        baseImage = 'java:openjdk-8-jre'
```
Comment (ahmetoz) [inline gradle-scripts/docker.gradle]: why not `java:openjdk-8-jre-alpine` ?

Comment (heshamMassoud) [reply gradle-scripts/docker.gradle]: b73bde5 thanks!

Actually I used it without thinking. Didn't know that the alpine was that much smaller. Esp. that we (don't, right?) need the extra OS stuff in the normal one.

File: docs/CONTRIBUTING.md (hunk 4 above)
Comment (heshamMassoud) [reply docs/CONTRIBUTING.md]: 46a12b0

File: build.gradle (hunk 2 above)
Comment (heshamMassoud) [reply build.gradle]: 7f2bb7c

**Pull request merged**